src/routes/challenge.rs

```diff
@@ -1,52 +1,70 @@
+use actix_session::Session;
 use actix_web::{get, post, web, HttpResponse, Scope};
+use rand::Rng;
 use sodiumoxide::crypto::box_;
-use uuid::Uuid;
 
 use crate::{
-    database::{
-        dto::{ChallengeData, ChallengeResponse},
-        querys::{generate_new_session, update_session},
-    },
+    database::dto::{ChallengeData, ChallengeResponse},
     error::AppError,
     AppState,
 };
 
 #[get("")]
-async fn challenge_handler(state: web::Data<AppState>) -> Result<HttpResponse, AppError> {
+async fn challenge_handler(
+    state: web::Data<AppState>,
+    session: Session,
+) -> Result<HttpResponse, AppError> {
     let app_name = state.app_name.clone();
-    let challenge = generate_new_session(&state.db_executor).await?;
 
-    let encryption_key_uri = state.session.key_uri.clone();
+    let mut rng = rand::thread_rng();
+    let challenge: [u8; 32] = rng.gen();
+
+    let encryption_key_uri = state.session_encryption_public_key_uri.clone();
     let challenge_data = ChallengeData {
-        challenge: challenge.as_bytes().to_vec(),
+        challenge: challenge.to_vec(),
         app_name,
         encryption_key_uri,
     };
+
+    session
+        .insert("challenge", challenge_data.clone())
+        .map_err(|_| AppError::Challenge("Could not insert encryption key"))?;
+
     Ok(HttpResponse::Ok().json(challenge_data))
 }
 
 #[post("")]
 async fn challenge_response_handler(
     state: web::Data<AppState>,
     challenge_response: web::Json<ChallengeResponse>,
+    session: Session,
 ) -> Result<HttpResponse, AppError> {
+    let challenge = session
+        .get::<ChallengeData>("challenge")
+        .map_err(|_| AppError::Challenge("Session not set"))?
+        .ok_or(AppError::Challenge("Session not set"))?
```

Comment:
double mapping of error

Reply:
I get an `option` after the `map_err` that's why the `ok_or` is called. Do you know a smooth way to directly transform it into `Result<String, AppError>`?